### PR TITLE
[openweathermap] Add daily moon channels to OneCall Thing

### PR DESCRIPTION
**Pull request by ErikDB87**
This adds the daily moon data provided by the API (https://openweathermap.org/api/one-call-api#parameter) to the OneCall Thing.

### CODE REVIEW

**Review by wborn — COMMENTED** [review]
Why does the PR also change other bundles?

* org.openhab.binding.mqtt.generic
* org.openhab.persistence.jpa

**Comment by ErikDB87** [comment]
> Why does the PR also change other bundles?
> 
> * org.openhab.binding.mqtt.generic
> * org.openhab.persistence.jpa

Mmm, that wasn't the plan... During the process (which was a first time), I did get the notification that meanwhile other requests had been pulled, so I had to update my clone. Maybe that's where something went wrong?

What can I do to fix this?

**Comment by ErikDB87** [comment]
> Maybe you can rebase it to fix it. It happens if you merge other changes into your branch as outlined in the pull request template:
> 
> > ATTENTION: Don't use "git merge" when working with your pull request branch!
> > This can clutter your Git history and make your PR unusable.
> > Use "git rebase" instead. See this forum post for further details:
> > https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

I tried something, and then entered command "gh pr edit" and chose body. Did that change anything?

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/OpenWeatherMapBindingConstants.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/OpenWeatherMapBindingConstants.java:92]: Please follow the naming convention used by this binding (and sidenote: required for new bindings: https://www.openhab.org/docs/developer/guidelines.html#naming-convention):

```suggestion
    public static final String CHANNEL_MOON_PHASE = "moon-phase";
```

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:80]: No need to add those setters to a DTO.
I wonder why the other fields have setters, probably because according to the JavaDoc a code generator was used.

Can you therefore please remove the added setters `setMoonrise`, `setMoonset`, `setMoon_phase`?

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:92]: Please follow the Java naming conventions (https://java.about.com/od/javasyntax/a/nameconventions.htm):

```suggestion
    public double getMoonPhase() {
```

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:38]: To follow the Java naming conventions, rename that field and add an annotation:

```suggestion
    @SerializedName("moon_phase")
    private double moonPhase;
```

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:93]:
```suggestion
        return moonPhase;
```

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java:676]:
```suggestion
                state = getDecimalTypeState(forecastData.getMoonPhase());
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap.properties
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap.properties:79]: Even though these lines are generated by the i18n tool, they are not needed, so please remove them after regenerating the i18n files (final step after changing any of the .xml files and to do before committing and pushing):

```suggestion
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap.properties
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap.properties:80]:
```suggestion
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:246]:
```suggestion
			<channel id="moon-phase" typeId="moon-phase"/>
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:284]:
```suggestion
			<channel id="moon-phase" typeId="moon-phase"/>
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:439]: It should be possible to add state description options, so openHAB automatically displays 'new moon' etc., but the periods in between could be an issue.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:432]:
```suggestion
	<channel-type id="moon-phase">
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:1]: Update instructions don't work that way, you need to define a new instruction set and increase the `thingTypeVersion` to 2 there: https://github.com/ErikDB87/openhab-addons-openweathermap/blob/d51cf0f2a1f6dcd27e8e640895fadcc3301c7cc5/bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/thing-types.xml#L278.

I am not sure how to the define update instructions for the many daily channel groups.
@openhab/add-ons-maintainers Do you know how to do that or if that is possible at all?

**Review by florian-h05 — CHANGES_REQUESTED** [review]
Thanks for your PR!

I have reviewed your changes, please have a look at my comments.
I tried to actually propose changes whereever possible, you should be able to click on an "add to batch" button for the proposed code changes and then finally commit all proposed changes in a single commit.

**Comment by florian-h05** [comment]
FTR: https://community.openhab.org/t/moon-information-from-one-call-openweathermap/153397

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (ErikDB87) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:439]: I was wondering about this, but decided to kick the can down the road :)

I do assume that there won't be periods in between the 4 stages. Won't it operate similar to a digital clock, which for example shows 8h43 for 60 seconds, until it's 8h44? (Similarly, I expect the moon-phase (MOON_PHASE? ;)) to be "0.25" for 7 days and then turn to "0.5".

But as to how that would show as "waning crescent" etc. in openHAB... No idea (yet).

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (ErikDB87) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:1]: I don't quite understand how this all works, but for now, I changed the version number to "2" in both documents.

**Comment by florian-h05** [comment]
Please do not force push after a review has been started (unless asked to because a rebase is needed), and please do not mark review comments as resolved, this the reviewers job.

Just letting you know this for the future 👍

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:1]: I will create a PR to your branch to fix the update instructions, please do not force push from now on.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:439]: I understand the API docs in a way that the moon phase float can be between 0.25 and 0.5, the closer you get to half moon, the close that float gets to 0.5.

State description options (https://www.openhab.org/docs/developer/bindings/thing-xml.html#state-description) only work for equality, so I don’t think they are very useful here.

Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:439]: You could however add an additional channel that does the “translation”.

**Comment by ErikDB87** [comment]
> Please do not force push after a review has been started (unless asked to because a rebase is needed), and please do not mark review comments as resolved, this the reviewers job.
> 
> Just letting you know this for the future 👍

Apologies!

What is the proper way then to "update" the files in the pull request to those I've amended on my local computer?

**Comment by florian-h05** [comment]
> What is the proper way then to "update" the files in the pull request to those I've amended on my local computer?

Just create a new commit and push that new commit.
Most IDE UIs have support for git, you don’t need to go to the command line to commit or push.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap.properties
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap.properties:360]: Remove this block here, it is not needed.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml
Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:439]: But I’d first focus on getting this PR finished before considering this.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:7]:
```suggestion
		<instruction-set targetVersion="1">
```

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:70]: Create a new instruction set after the existing one and put these three instructions into it (please fix the formatting with `mvn spotless:apply -pl :org.openhab.binding.openweathermap`):

```xml
<instruction-set targetVersion="2">
    <add-channel id="moonrise" groupIds="forecastDaily">
        <type>openweathermap:moonrise</type>
    </add-channel>
			<add-channel id="moonset" groupIds="forecastDaily">
				<type>openweathermap:moonset</type>
			</add-channel>
			<add-channel id="moon_phase" groupIds="forecastDaily">
				<type>openweathermap:moon_phase</type>
			</add-channel>
</instruction-set>
```

**Comment by ErikDB87** [comment]
> > What is the proper way then to "update" the files in the pull request to those I've amended on my local computer?
> 
> Just create a new commit and push that new commit. Most IDE UIs have support for git, you don’t need to go to the command line to commit or push.

Ah. I'll try to familiarize myself with the Visual Studio Code plugin then. Although I don't dislike working with CLI's :)

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (ErikDB87) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:1]: > I will create a PR to your branch to fix the update instructions, please do not force push from now on.

Only now saw this reply. So I'll sit back and let you do all the rest? :)

Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:1]: I noticed I can’t create a PR from GitHub web because of the web code editor is not that great and I’m not near my laptop, but I have commented the required changes down below.

**Comment by florian-h05** [comment]
> Ah. I'll try to familiarize myself with the Visual Studio Code plugin then. Although I don't dislike working with CLI's :)

CLIs sometimes are great IMO, but as I know the important keyboard shortcuts in IntelliJ, for me it’s faster to do the common stuff and for the stuff I don’t need that often, it is helpful to not have to remember the actual git commands.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (florian-h05) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:1]: One minor comment about codestyle:

Since moon data is coming after sun data in the API response and the class field order, the moon data getters should come after the sun data methods, i.e. move the new methods after `setSunset`.

**Review by florian-h05 — CHANGES_REQUESTED** [review]
LGTM, thanks!
Just one final code style comment.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (ErikDB87) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:1]: Sorry, I'm stuck. I made the changes in Visual Studio Code, and tried commiting through the Github UI in there. But then the commits weren't signed off, so the "DCO" test failed. And every time I try to rebase to a previous commit, I get an error if I don't pull. But if I pull, the latest changes are already done (because of what I did in Visual Studio Code, I assume?). So I can't get my commit signed.

I'm caught in a loop.

Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:1]: The DCO bot tells how to fix that: https://github.com/openhab/openhab-addons/pull/16350/checks?check_run_id=21170803740

But you don't need to. Since you signed-off the other commits and forgot it by accident on 2 commits, I will just tell the DCO bot to accept that.

**Review by florian-h05 — APPROVED** [review]
LGTM, thanks!

@openhab/add-ons-maintainers Can you please review (or alternatively trust my review) and merge this PR?

**Comment by florian-h05** [comment]
@ErikDB87 Just one thing missing: Can you please update the README to mention the channels?
Just add it to the table here https://github.com/ErikDB87/openhab-addons-openweathermap/blob/f1effe9387967f66f3c64be13a020c36efd575df/bundles/org.openhab.binding.openweathermap/README.md?plain=1#L219 and after that fix the formatting of the table with http://markdowntable.com or your IDE.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
Comment (ErikDB87) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java:1]: I tried `git rebase`, but `Current branch main is up to date`.
I tried `git rebase HEAD~13 --signoff`, but:
```
Auto-merging bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
CONFLICT (content): Merge conflict in bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
error: could not apply 6d5bd47c72... Final commit?
```

So you'll have to work your magic again, I'm afraid... :|

**Comment by ErikDB87** [comment]
> @ErikDB87 Just one thing missing: Can you please update the README to mention the channels? Just add it to the table here https://github.com/ErikDB87/openhab-addons-openweathermap/blob/f1effe9387967f66f3c64be13a020c36efd575df/bundles/org.openhab.binding.openweathermap/README.md?plain=1#L219 and after that fix the formatting of the table with http://markdowntable.com or your IDE.

I tried `git rebase`, but `Current branch main is up to date`.
I tried `git rebase HEAD~13 --signoff`, but:
```
Auto-merging bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
CONFLICT (content): Merge conflict in bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Daily.java
error: could not apply 6d5bd47c72... Final commit?
```

So you'll have to work your magic again, I'm afraid... :|

**Comment by florian-h05** [comment]
@ErikDB87 I have fixed the formatting in the README and set DCO to pass.

@jlaur Can you please have a look and merge this PR?

**Comment by ErikDB87** [comment]
> @ErikDB87 I have fixed the formatting in the README and set DCO to pass.
> 
> @jlaur Can you please have a look and merge this PR?

Thanks! I didn't fully understand the formatting issue :|

**Comment by lsiepel** [comment]
Can you fix the conflict? I might have some time to review the next days as it is a holiday week here :-)

**Comment by florian-h05** [comment]
@lsiepel Done.
FYI I did a local rebase and force push after GitHub web resolve did a merge commit which cluttered the Git history a bit.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (jlaur) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:155]: I think group id `forecastToday` should be added since the channels were also added to channel group type id `oneCallDaily` referenced by that group?
```suggestion
			<add-channel id="moonrise" groupIds="forecastToday,forecastDaily">
				<type>openweathermap:moonrise</type>
			</add-channel>
			<add-channel id="moonset" groupIds="forecastToday,forecastDaily">
				<type>openweathermap:moonset</type>
			</add-channel>
			<add-channel id="moon-phase" groupIds="forecastToday,forecastDaily">
				<type>openweathermap:moon-phase</type>
			</add-channel>
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thank you! Only one question regarding the update instructions.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml
Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:155]: Yes. Shouldn't those also be added to `forecastTomorrow`, `forecastDay2` ... `forecastDay7`?

Comment (jlaur) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:155]: Oh, right, I missed those. 🙄

Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml:155]: I will add them, I've already helped with this PR.

Comment (florian-h05) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/update/instructions.xml]: Done :+1:

**Review by jlaur — APPROVED** [review]
LGTM. @lsiepel - did you start a review also during your holiday? 🙂 Feel free to merge when you are fine with it also or if you'd like to skip.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**
